captain/parser/bool_parser/bool_parser.py

```diff
@@ -0,0 +1,250 @@
+                    if match_item not in cur_node.children:
+                        cur_node.children[match_item] = TrieNode(match_item, {}, False)
+                    cur_node = cur_node.children[match_item]
+                cur_node.is_end = True
+            return root
+
+        def is_match(
+            cur_idx: int, trie: TrieNode
+        ) -> tuple[bool, list[Expression], Type[Expression]]:
+            cur_node = trie
+            targets = []
+            expr_type: Type[Expression] = None  # type: ignore
+            try:
+                while not cur_node.is_end:
+                    cur_item = to_parse[cur_idx]
+                    item_matcher = (
+                        cur_item.returns
+                        if isinstance(cur_item, Expression)
+                        else cur_item
+                    )
+
+                    if item_matcher not in cur_node.children:
+                        return False, None, None  # type: ignore
+                    cur_node = cur_node.children[item_matcher]
+                    cur_idx += 1
+                    if isinstance(cur_item, Operator):
+                        expr_type = TOKEN_TO_CLASS[cur_item.operator]
+                    else:
+                        targets.append(cur_item)
+
+                if not expr_type:
+                    raise MatchError(
+                        "could not find the expression type (was an operator provided?)"
+                    )
+                return True, targets, expr_type
+            except IndexError:
+                return False, None, None  # type: ignore
+
+        trie = _build_trie(to_match)
+        for i in range(0, len(to_parse)):
+            result, targets, expr_type = is_match(i, trie)
+            if result:
+                expression = expr_type(targets)
+                return result, i, i + len(expr_type.expects), expression
+
+        return False, None, None, None  # type: ignore
+
+    """
+    ____________________________
+    """
+    to_parse: List[ParseItem] = []
+
+    # step 1: parse all subexpressions first
+    sub_exps = (
+        [(-2, -1)] + _get_subexpressions(tokens) + [(len(tokens), len(tokens) + 1)]
+    )
+    i = -2
+    for sub_exp in sub_exps:
+        to_parse += tokens[i : sub_exp[0]]
+        sublist = tokens[sub_exp[0] + 1 : sub_exp[1]]
+        if sublist:
+            to_parse += [recurse(sublist)]
+        i = sub_exp[1] + 1
+
+    # step 2: parse in order defined by parser_config
+    for expressions in rules.order_of_operations:
+        matched, match_idx, end_idx, expr = _match(to_parse, expressions)
+        while matched:
+            to_parse = to_parse[:match_idx] + [expr] + to_parse[end_idx:]
+            matched, match_idx, end_idx, expr = _match(to_parse, expressions)
+    if len(to_parse) != 1 or not isinstance(to_parse[0], Expression):
+        raise InvalidExpression(f"At the end of parse, got: {to_parse}")
+
+    return to_parse[0]
```

Comment:
this entire function is way way way too big, break it into chunks to make things easier to understand

Reply:
are you referring to _match() ? I can make it smaller, but I already have like 4 helper functions inside to make it more readable.

Reply:
Also really not sure about nesting a class inside of a method, even if the class is only used by this method.